File: 백준/Gold/2133. 타일 채우기/타일 채우기.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static StringTokenizer st;
    private static StringBuilder sb = new StringBuilder();
    private static int N;
    private static int[] dp;

    public static void main(String[] args) throws Exception {
        N = Integer.parseInt(br.readLine());
        if (N % 2 != 0) {
            System.out.println(0);
            return;
        }
        dp = new int[N + 1];
        dp[0] = 1;
        dp[2] = 3;
        for (int i = 4; i <= N; i += 2) {
            dp[i] = dp[i - 2] * 3;
            for (int j = 4; j <= i; j += 2) {
                dp[i] += (dp[i - j] * 2);
            }
        }
        System.out.println(dp[N]);
    }

}

File: 백준/Gold/2133. 타일 채우기/타일 채우기.kt
import java.io.BufferedReader
import java.io.InputStreamReader


fun main(args: Array<String>) {

    val br = BufferedReader(InputStreamReader(System.`in`))
    val N = br.readLine().toInt()

    if (N % 2 != 0) {
        println(0)
        return
    }

    val dp = IntArray(N + 1)
    dp[0] = 1;
    dp[2] = 3;
    for (i in 4..N step 2) {
        dp[i] = dp[i - 2] * 3
        for (j in 4..N step 2) {
            if (i - j >= 0) {
                dp[i] += (dp[i - j] * 2)
            }
        }
    }

    println(dp[N])
}